### ISSUE TITLE
Erasure channel populates the DEM in an unexpected way

### ISSUE THREAD

**Issue by markturner289** [quantumlib/Stim#602]
Exciting to see a heralded erasure channel added into Stim! I had a play with it but am wondering whether it populates the DEM as we would expect.

I'm looking at the following circuit snippet where erasure is certain:

```
HERALDED_ERASE(1.0) 0
M 0
DETECTOR rec[-1]
```

It yields the following DEM:

```
error(0.375) D0
```

Since the erasure channel maximally mixes the target qubit and the erasure probability in the circuit snippet is 100%, I would have expected the DEM to read `error(0.5) D0`. Indeed, this is exactly the DEM I see with the following circuit I understand to be equivalent to the above bar the fact the error cannot be heralded.

```
DEPOLARIZE1(0.75) 0
M 0
DETECTOR rec[-1]
```

If this is a bug, I'd be keen to propose a patch upstream to fix.

**Comment by Strilanc** [comment]
So, this is a bug, but it's not the bug you think it is. Things are clearer if you also declare the herald as a detector:

```
import stim

print(stim.Circuit("""
    HERALDED_ERASE(1.0) 0
    M 0
    DETECTOR rec[-1]
    DETECTOR rec[-2]
""").detector_error_model())
```

```
error(0.375) D0 D1
error(0.375) D1
```

Here `D0` is the measurement being wrong and `D1` is the erasure herald going off. The property you're talking about in your issue is that P(D0|D1) should be 0.5 and in fact here it is 0.5 as required. The confusing thing is instead that P(D1) is not 100%.

The underlying issue is that detector error models can only represent independent error mechanisms, and so there's actually no way to write down "the herald goes off and you project to the no error case OR ELSE DISJOINTLY the herald goes off and you project to the yes error case".

The fact it picks 0.375 as the approximation... I'll look into that. That strikes me as not the best approximation for this case.

The fact is uses an approximation without requiring the user to set `allow_approximations=True`: that's a bug and I will fix it.

The fact that DEMs can't say the required thing: I also intend to fix this, by adding a `oneof {}` block and an `if D {}` block to the specification.

**Comment by Strilanc** [comment]
The goal of the heralded erasure channel isn't to be exactly like a
specific physical process, it's to provide a mechanism that's easy for the
simulator to do and easy to mathematically understand that can be used to
roughly approximate the effects of those channels. If the erasure that
you're trying to model can't happen twice in one round, the way you would
deal with that is to not put it in the circuit twice in one round.

Basically, in the same way that you approximate t1 decay as a probabilistic
bit flip (despite the fact that the bit flip is self inverse instead of
idempotent), you approximate losing a photon as a heralded 50/50 bit flip
and phase flip.
